File: streamer_scripts/ipc_manager.py
from posix_shmem import PosixShmem
from posix_mqueue import PosixMQueue
from posix_semaphore import PosixSemaphore

import cv2
import numpy as np

SHMEM_DEFAULT_SIZE = 2500 * 2000 * 3


class IpcManager:

    def __init__(self, shmem_name, sem_name, mq_name):
        self._shmem = PosixShmem(shmem_name, SHMEM_DEFAULT_SIZE)
        self._sem = PosixSemaphore(sem_name)
        self._mq = PosixMQueue(mq_name)

    def __str__(self):
        return "StreamerIpcManager"

    def __call__(self, *args, **kwargs):
        pass

    def write_frame(self, frame):
        assert type(frame) == np.ndarray
        ysz, xsz, csz = frame.shape
        msg = str(ysz * xsz * csz) + " " + str(16) + " " + str(ysz) + " " + str(xsz)
        if self._sem.lock(timeout=0.010):
            self._shmem.write(frame)
            self._sem.unlock()
            self._mq.send(msg)

    def read_frame(self):
        msg = self._mq.receive(timeout=1.0)
        if msg is None:
            return None
        msg = msg[:-1].decode("utf-8")
        self._sem.lock(timeout=0.010)
        args = msg.split()
        total, img_type, ysz, xsz = int(args[0]), int(args[1]), int(args[2]), int(args[3])
        data = self._shmem.read()
        img = data[:total].copy()
        img = img.reshape((ysz, xsz, 3))
        self._sem.unlock()
        return img

File: compression_streamer/ipc_manager.cpp
#include "ipc_manager.h"

#include <sstream>

namespace ipc {

    IpcManager::IpcManager(std::string shmem_name, std::string sem_name, std::string mq_name)
        : _shm(shmem_name, ShmemDedfaultSize),
        _sem(sem_name),
        _mq(mq_name) {
    }

    void IpcManager::write_frame(const cv::Mat& img) const {
        size_t frame_size = img.total() * img.elemSize();
        auto shmem_ptr = _shm.get_pointer();
        std::stringstream ss;
        ss << frame_size << " " << img.type() << " " << img.rows << " " << img.cols;
        _sem.lock();
        std::copy(img.data, img.data + frame_size, shmem_ptr);
        _sem.unlock();
        _mq.send(ss.str());
    }

    cv::Mat IpcManager::read_frame() const {
        std::string msg = _mq.receive();
        _sem.lock();
        std::stringstream ss(msg);
        size_t imgsz;
        int type, rows, cols;
        ss >> imgsz >> type >> rows >> cols;
        cv::Mat img(rows, cols, type, cv::Scalar(0));
        auto ptr = _shm.get_pointer();
        std::copy(ptr, ptr + imgsz, img.data);
        _sem.unlock();
        return img;
    }

    void operator>>(const cv::Mat& img, IpcManager& ipc) {
        ipc.write_frame(img);
    }

    void operator>>(IpcManager& ipc, cv::Mat& img) {
        img = ipc.read_frame();
    }

}
